Add inference_params support for LLM Judges

debu-sinha opened the pull request:
### Related Issues/PRs

Resolves #19120

### What changes are proposed in this pull request?

This PR exposes inference parameters (e.g., temperature, top_p, max_tokens) for LLM Judges, enabling more deterministic and controllable evaluation results. This was a feature that existed in MLflow < 3 and is being re-introduced based on user feedback.

**Changes:**
- Add `inference_params` parameter to `make_judge()` function
- Add `inference_params` to `InstructionsJudge` class with property accessor
- Pass `inference_params` through `invoke_judge_model` to LiteLLM, Databricks serving endpoint, and gateway adapters
- Update `get_chat_completions_with_structured_output` to accept `inference_params`
- Add serialization support for `inference_params` in `model_dump`
- Include `inference_params` in `__repr__` when set
- Add FAQ section "How do I pass additional inference parameters to judge LLMs?"

**Example usage:**
```python
from mlflow.genai import make_judge

# Create a judge with deterministic outputs (temperature=0)
deterministic_judge = make_judge(
    name="quality_check",
    instructions="Evaluate if {{ outputs }} is accurate and provide a rationale.",
    model="openai:/gpt-4o-mini",
    inference_params={"temperature": 0.0},
)

# Create a judge with varied outputs (temperature=1)
varied_judge = make_judge(
    name="quality_check",
    instructions="Evaluate if {{ outputs }} is accurate and provide a rationale.",
    model="openai:/gpt-4o-mini",
    inference_params={"temperature": 1.0},
)

test_output = {"answer": "The Eiffel Tower is 330 meters tall."}
```

**Actual output with `temperature=0.0` (deterministic - identical rationales):**
```
Run 1: value=accurate
       rationale=The statement that the Eiffel Tower is 330 meters tall is accurate. 
       The height of the Eiffel Tower, including its antennas, is indeed approximately 
       330 meters. This measurement is widely accepted and verified by various sources, 
       making the output correct.

Run 2: value=accurate
       rationale=The statement that the Eiffel Tower is 330 meters tall is accurate. 
       The height of the Eiffel Tower, including its antennas, is indeed approximately 
       330 meters. This measurement is widely accepted and verified by various sources, 
       making the output correct.
```

**Actual output with `temperature=1.0` (varied - different rationales each run):**
```
Run 1: value=accurate
       rationale=The height of the Eiffel Tower is indeed approximately 330 meters, 
       including its antennas. This information is precise and aligns with reliable 
       data regarding the structure's height, thus making the answer accurate.

Run 2: value=accurate
       rationale=The output provides a precise measurement of the height of the Eiffel 
       Tower, which is indeed 330 meters tall including its antennas. This information 
       is factually correct and aligns with common knowledge about the landmark.

Run 3: value=accurate
       rationale=The height of the Eiffel Tower is indeed approximately 330 meters, 
       including the antennas. This measurement is well-documented and widely accepted, 
       making the output correct regarding the height of the Eiffel Tower.
```

### How is this PR tested?

- [x] New unit/integration tests

Tests cover:
- `make_judge()` with and without `inference_params` (parameterized)
- Passing `inference_params` through to `invoke_judge_model`
- `inference_params` in `__repr__` and serialization
- LiteLLM adapter with various parameter combinations
- Databricks serving endpoint adapter with `inference_params`
- Tool calling loop with persistent `inference_params`

### Does this PR require documentation update?

- [x] Yes. Added FAQ section in `docs/docs/genai/eval-monitor/faq.mdx`

### Release Notes

#### Is this a user-facing change?

- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Add `inference_params` parameter to `make_judge()` and `InstructionsJudge`, allowing users to configure LLM inference parameters (e.g., `temperature`, `top_p`, `max_tokens`) for more deterministic and reproducible evaluation results.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes

#### Should this PR be included in the next patch release?

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

debu-sinha committed:
Add inference_params support for LLM Judges

This change exposes inference parameters (e.g., temperature, top_p, max_tokens)
for LLM Judges, enabling more deterministic and controllable evaluation results.

Changes:
- Add inference_params parameter to make_judge() function
- Add inference_params to InstructionsJudge class with property accessor
- Pass inference_params through invoke_judge_model to litellm.completion
- Update get_chat_completions_with_structured_output to accept inference_params
- Add serialization support for inference_params in model_dump
- Include inference_params in __repr__ when set

Closes #19120

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

B-Step62 commented on tests/genai/judges/test_make_judge.py:
```diff
@@ -3509,3 +3511,130 @@ def test_response_format_uses_generic_field_description(description):
     assert result_field.description == _RESULT_FIELD_DESCRIPTION
+
+
+class TestInferenceParams:
```
We don't use classes in unit tests, please follow the consistent style as other tests.

B-Step62 reviewed (commented):
Overall looks good, but I think we should make the behavior of other type of judges like built-in, non-litellm, databricks, consistent. cc: @AveshCSingh

debu-sinha committed:
Address review feedback: refactor tests and add consistent inference_params support

Changes based on reviewer feedback from @B-Step62:
- Convert class-based tests to function-based tests following MLflow testing conventions
- Remove single-line docstrings from test functions per MLflow style guidelines
- Add inference_params support to Databricks serving endpoint adapter for consistency
- Add inference_params support to gateway adapter (non-litellm fallback) for consistency

This ensures inference_params works consistently across all judge model backends:
- LiteLLM adapter (already supported)
- Databricks serving endpoint adapter (now supported)
- AI Gateway adapter (now supported)

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

smoorjani commented on tests/genai/judges/test_make_judge.py:
```diff
@@ -3509,3 +3514,126 @@ def test_response_format_uses_generic_field_description(description):
+    assert judge.inference_params == inference_params
+
+
+def test_inference_params_passed_to_invoke_judge_model(mock_invoke_judge_model):
```
can we merge this test with the first one? same with `test_inference_params_in_repr` with `test_make_judge_with_inference_params` and `test_inference_params_not_in_repr_when_none` with `test_make_judge_without_inference_params`. same with the serialization tests too

smoorjani commented on tests/genai/judges/test_make_judge.py:
```diff
@@ -3509,3 +3514,126 @@ def test_response_format_uses_generic_field_description(description):
+    assert "inference_params" not in pydantic_data
+
+
+@pytest.mark.parametrize(
```
maybe let's just parameterize the first test with these?

smoorjani commented on tests/genai/judges/utils/test_invocation_utils.py:
```diff
@@ -984,3 +984,139 @@ class FieldExtraction(BaseModel):
+    assert "temperature" not in call_kwargs
+
+
+def test_inference_params_with_temperature_zero(mock_response):
```
redundant with first test in this file?

smoorjani commented on tests/genai/judges/utils/test_invocation_utils.py:
```diff
@@ -984,3 +984,139 @@ class FieldExtraction(BaseModel):
+
+    with mock.patch("litellm.completion", return_value=mock_response) as mock_litellm:
+        invoke_judge_model(
+            model_uri="openai:/gpt-4",
```
can we unit test the path with a databricks serving endpoint as well?

smoorjani reviewed (approved):
looks great! just a few nits on the tests to address before merging. 

also, can you include the result of the example code in the PR description? maybe run it twice and then show how running twice with `temperature=1.0` results in different outputs

debu-sinha committed:
Address review feedback: refactor tests and add Databricks endpoint test

- Convert to function-based tests (no class-based tests)
- Consolidate redundant tests using parameterization
- Remove test_instructions_judge_with_inference_params (non-public interface)
- Merge repr and serialization tests into main test functions
- Add test_invoke_databricks_serving_endpoint_with_inference_params

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha commented:
Updated the PR description with **actual test outputs** demonstrating the feature:

- `temperature=0.0`: Produces identical rationales across runs (deterministic)
- `temperature=1.0`: Produces varied rationales each run (different wording, same conclusion)

The outputs were generated by running the example code with a real OpenAI API key.

debu-sinha commented:
@smoorjani Thanks for the approval! All feedback has been addressed - tests refactored, Databricks endpoint test added, and PR description updated with actual test outputs. Is this ready to merge?

debu-sinha committed:
Add Databricks demo and test screenshots for inference_params feature

Adds a demo notebook and test screenshots showing the inference_params
feature working on Databricks:

- databricks_demo.py: Databricks notebook demonstrating all test cases
- Screenshots from successful test runs on Databricks

Test coverage:
1. Deterministic outputs with temperature=0.0
2. Varied outputs with temperature=1.0
3. Multiple inference parameters (temperature, max_tokens, top_p)
4. Default behavior without inference_params
5. Integration with mlflow.genai.evaluate()

Model tested: databricks-claude-sonnet-4

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Add OSS test evidence for inference_params feature

Adds OSS (LiteLLM/OpenAI) test script and output demonstrating the
inference_params feature works in non-Databricks environments.

Test results:
- Test 1 (temp=0.0): All 3 rationales identical - PASS
- Test 2 (temp=1.0): Rationales varied across runs - PASS
- Test 3 (multiple params): All params applied - PASS
- Test 4 (default): inference_params=None - PASS

Model tested: openai:/gpt-4o-mini via LiteLLM adapter

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Remove test artifacts from code diff

Move manual test screenshots and notebooks to PR comments per reviewer request.
The test evidence is preserved in the PR discussion.

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha commented:
Removed the test artifacts from the code diff as requested. The screenshots are backed up locally and I'll upload them directly to a PR comment shortly.

@B-Step62 - To confirm, should I add the screenshots as images in a new comment here, or would you prefer them embedded in the PR description?

debu-sinha committed:
Add FAQ section for passing inference parameters to judge LLMs

Address reviewer feedback by adding documentation to FAQ instead of
example code in the repository. Documents the inference_params argument
for LLM-as-a-Judge scorers like Correctness and Guidelines.

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

B-Step62 reviewed (approved):
LGTM!

debu-sinha committed:
Fix blacken-docs formatting in FAQ

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Fix inference_params field missing from AdapterInvocationInput

Add inference_params field to AdapterInvocationInput dataclass to fix
TypeError when passing inference parameters through the judge model
invocation chain. Also propagate inference_params through LiteLLM and
Databricks serving endpoint adapters. Update tests to expect the new
parameter.

Signed-off-by: Debu Sinha <debu.sinha@databricks.com>
Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Merge branch 'master' into feature/expose-inference-params-for-llm-judges
